File: python-alert-service/clients/sensor_client.py
"""Sensor service client with circuit breaker and retry logic."""

import logging
import time

import httpx
import pybreaker

logger = logging.getLogger("alert_service")


class _LogListener(pybreaker.CircuitBreakerListener):
    """Logs circuit breaker state transitions."""

    def state_change(self, cb, old_state, new_state):
        logger.warning(
            "Circuit breaker state changed name=%s from=%s to=%s",
            cb.name,
            str(old_state),
            str(new_state),
        )


class SensorClient:
    """HTTP client for the sensor service with resilience patterns.

    Uses a circuit breaker (pybreaker) to prevent cascade failures.
    Retry with exponential backoff is applied *inside* the circuit breaker
    call so the entire retry sequence counts as one failure toward fail_max.
    Falls back gracefully when the sensor service is unavailable.
    """

    def __init__(
        self,
        base_url: str,
        api_token: str,
        cb_fail_max: int = 5,
        cb_reset_timeout: int = 30,
    ):
        self.base_url = base_url
        self.api_token = api_token
        self.breaker = pybreaker.CircuitBreaker(
            fail_max=cb_fail_max,
            reset_timeout=cb_reset_timeout,
            name="sensor-service",
            listeners=[_LogListener()],
        )

    def _make_request_with_retry(self, sensor_id: str, max_retries: int = 3) -> dict | None:
        """Make HTTP request with retry and exponential backoff.

        Runs inside the circuit breaker call so the full retry sequence
        counts as one attempt toward fail_max. Backoff: 1s, 2s between attempts.
        Timeout per request: 2 seconds.
        """
        last_err: Exception | None = None
        for attempt in range(max_retries):
            backoff = 0
            if attempt > 0:
                backoff = 2 ** (attempt - 1)  # 1s, 2s
                time.sleep(backoff)
            try:
                response = httpx.get(
                    f"{self.base_url}/sensors/{sensor_id}",
                    headers={"Authorization": f"Bearer {self.api_token}"},
                    timeout=2.0,
                )
                if response.status_code == 404:
                    return None
                response.raise_for_status()
                return response.json()
            except Exception as e:
                last_err = e
                logger.warning(
                    "Sensor service request failed, retrying sensor_id=%s attempt=%d max_retries=%d backoff_ms=%d error=%s",
                    sensor_id,
                    attempt + 1,
                    max_retries,
                    backoff * 1000,
                    str(e),
                )
        logger.warning(
            "All retries exhausted for sensor service sensor_id=%s max_retries=%d error=%s",
            sensor_id,
            max_retries,
            str(last_err),
        )
        raise last_err

    def get_sensor(self, sensor_id: str) -> tuple[dict | None, bool]:
        """Get sensor data via circuit breaker.

        Returns:
            A tuple of (sensor_data, is_validated) with three possible outcomes:
            - (sensor_data, True): Sensor found and validated successfully
            - (None, True): Sensor confirmed not found (404 from sensor service)
            - (None, False): Sensor service unavailable — fallback, not validated
        """
        try:
            result = self.breaker.call(self._make_request_with_retry, sensor_id)
            return result, True
        except pybreaker.CircuitBreakerError:
            logger.warning(
                "Circuit breaker is OPEN — sensor service unavailable sensor_id=%s",
                sensor_id,
            )
            return None, False
        except Exception as e:
            logger.warning(
                "Sensor service unavailable sensor_id=%s error=%s",
                sensor_id,
                str(e),
            )
            return None, False

File: go-alert-service/clients/sensor_client.go
package clients

import (
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"math"
	"net/http"
	"time"

	"github.com/sony/gobreaker"
)

// SensorClient is an HTTP client for the sensor service with circuit breaker
// and retry logic for resilient inter-service communication.
type SensorClient struct {
	baseURL  string
	apiToken string
	cb       *gobreaker.CircuitBreaker
	client   *http.Client
}

// NewSensorClient creates a new sensor client with circuit breaker.
func NewSensorClient(baseURL, apiToken string, failMax int, resetTimeout int) *SensorClient {
	settings := gobreaker.Settings{
		Name:        "sensor-service",
		MaxRequests: 1,
		Interval:    time.Duration(resetTimeout) * time.Second,
		Timeout:     time.Duration(resetTimeout) * time.Second,
		ReadyToTrip: func(counts gobreaker.Counts) bool {
			return counts.ConsecutiveFailures >= uint32(failMax)
		},
		OnStateChange: func(name string, from gobreaker.State, to gobreaker.State) {
			slog.Warn("Circuit breaker state changed",
				"name", name, "from", from.String(), "to", to.String())
		},
	}

	return &SensorClient{
		baseURL:  baseURL,
		apiToken: apiToken,
		cb:       gobreaker.NewCircuitBreaker(settings),
		client:   &http.Client{Timeout: 2 * time.Second},
	}
}

// GetSensor retrieves sensor data via the circuit breaker.
// Returns (sensor_data, is_validated, error):
//   - (data, true, nil): Sensor found and validated
//   - (nil, true, error): Sensor confirmed not found (404)
//   - (nil, false, nil): Sensor service unavailable (fallback)
func (sc *SensorClient) GetSensor(sensorID string) (map[string]interface{}, bool, error) {
	result, err := sc.cb.Execute(func() (interface{}, error) {
		return sc.makeRequestWithRetry(sensorID, 3)
	})

	if err != nil {
		slog.Warn("Sensor service unavailable",
			"sensor_id", sensorID, "error", err.Error())
		return nil, false, nil
	}

	if result == nil {
		return nil, true, fmt.Errorf("sensor not found")
	}

	return result.(map[string]interface{}), true, nil
}

// makeRequestWithRetry makes an HTTP request with retry and exponential backoff.
func (sc *SensorClient) makeRequestWithRetry(sensorID string, maxRetries int) (interface{}, error) {
	var lastErr error

	for attempt := 0; attempt < maxRetries; attempt++ {
		var backoff time.Duration
		if attempt > 0 {
			backoff = time.Duration(math.Pow(2, float64(attempt-1))) * time.Second
			time.Sleep(backoff)
		}

		result, err := sc.makeRequest(sensorID)
		if err == nil {
			return result, nil
		}
		lastErr = err
		slog.Warn("Sensor service request failed, retrying",
			"sensor_id", sensorID, "attempt", attempt+1, "max_retries", maxRetries,
			"backoff_ms", backoff.Milliseconds(), "error", err.Error())
	}

	slog.Warn("All retries exhausted for sensor service",
		"sensor_id", sensorID, "max_retries", maxRetries, "error", lastErr.Error())
	return nil, lastErr
}

// makeRequest makes a single HTTP request to the sensor service.
func (sc *SensorClient) makeRequest(sensorID string) (interface{}, error) {
	url := fmt.Sprintf("%s/sensors/%s", sc.baseURL, sensorID)

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+sc.apiToken)

	resp, err := sc.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return nil, nil
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("sensor service returned status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	return result, nil
}
